syntax="proto3";
option java_multiple_files = true;
option java_package = "com.diamond.diamond.grpc_client"; // Explicit package declaration

package com.diamond.diamond.grpc_client;

enum Blockchain {
    SOL = 0;
    SOL_DEVNET = 1;
    SOL_TESTNET = 2;
    EVM = 3;
}
/* Wallet Sets */

/*
Contains an idempotency key
*/
message CreateWalletSetRequest {
    optional string walletName = 1;
    //string idempotencyKey = 2;
}

message CreateWalletSetResponse {
    string id = 1;
}

/* Wallets */

message CreateWalletRequest {
    Blockchain blockchain = 1;
    string walletSetId = 2;
    string idempotencyKey = 3;
}

message CreateWalletsRequest {
    int32 count = 1;
    repeated Blockchain blockchains = 2;
    string walletSetId = 3;
    string idempotencyKey = 4;
}

message CreateWalletResponse {
    string id = 1;
    Blockchain blockchain = 2;
    string address = 3;
    string createDate = 4;
    string custodyType = 5;
    string name = 6;
    string walletSetId = 7;
    string accountId = 8;
    string initialPublicKey = 9;
}

message CreateWalletsResponse {
    repeated CreateWalletResponse wallets = 1;
}

service Circle {
    rpc CreateWalletSet (CreateWalletSetRequest) returns (CreateWalletSetResponse) {};
    rpc CreateWallets (CreateWalletsRequest) returns (CreateWalletsResponse) {}; // creating multiple wallets at once
    rpc CreateWallet (CreateWalletRequest) returns (CreateWalletResponse) {}; // creating only one wallet
}